### DISCUSSION TITLE
spacemouse support

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1205] [Ideas]
I have written a proof of concept spacemouse (a 6dof mouse/joystick device) support to VulkanSceneGraph through the spnav driver on Linux. I did it by simply interpreting the spnav events and handing them over to `trackball`. It all boils down to the following simple code in the main loop (e.g. of vsgviewer):

```c++
    double dx = 0, dy = 0, dz = 0, xrot = 0, yrot = 0, zrot = 0;

    spnav_event event;
    if ( spnav_poll_event(&event) > 0)
    {
      if (event.type == SPNAV_EVENT_MOTION)
      {
        dx = 0.0005* event.motion.x;
        dy = 0.0005* event.motion.y;
        dz = 0.0005* event.motion.z;
        xrot = -0.0005* event.motion.rx;
        yrot = -0.0005* event.motion.ry;
        zrot = -0.0005* event.motion.rz;
      }
      if (swapyz)
      {
        std::swap(dy, dz);
        dy = -dy;
        std::swap(yrot, zrot);
      }

      // Don't accumulate events
      spnav_remove_events(SPNAV_EVENT_MOTION);
    }

    trackball->pan({dx,dy});
    trackball->zoom(dz);
    trackball->rotate(xrot, vsg::dvec3(1,0,0));
    trackball->rotate(yrot, vsg::dvec3(0,1,0));
    trackball->rotate(zrot, vsg::dvec3(0,0,1));

```

I'm wondering: Is there any general interest for this code? If so, I need some instructions for how wrap these events "idiomatically". I thought of using a `Visitor`, but it obviously doesn't support the special events from the spnav library.

**Comment by robertosfield** [comment]
I'm not familiar with spnav dirver.  Could you provide more details about what libraries you are linking to, and hardware you are using?

I do have a gampad style device that I've used with the OSG many yars ago.  I used SDL to read from it's joysticks and buttons but didn't attempt to directly push events to the OSG's event handling.

Would could take the same approach with the VSG, just leaving how to integrate to 3rd party applications/libs.

Another route, would be extending the VSG's event handling is do via a combination of a vsg::UIEvent base class and vsg::Visitor/ConstVisitor to handle the various subclasses from UIEvent.   User applications/libraries could subclass from UIEvent but would need to handle the test for the custom UIEvent in the Custom Visitor::apply(EIEvent&).  

Finally, modifying the VSG itself to add a new UIEvent subclass to handle more sophisticated devices that pointer devices like a mouse/touch/keyboard would open the door to having the Visitor/ConstVisitor classes supported the new event type making it easier to catch those events.

I don't have any strong feelings in any direction at this point, the above approaches could do tackle in the above order, as the implementation and experience with it matures closer integration into the core VSG might be appropriate.  However, I don't want to start adding new dependencies to the VSG itself, so if we need to link to 3rd party libraries this would be best done via an additional library.  It might be that just a vsgExample would suffice if it's straight forward to integrate.

**Comment by dov** [comment]
Here's a link to my device:

https://3dconnexion.com/il/product/spacemouse-wireless/

After investigating how FreeCAD supports the device, I found that it interfaces with it through libspnav. It may be found here:

https://github.com/FreeSpacenav/libspnav 

Thanks for your thoughts about how to do the integration. I'm still not following everything you're suggesting though. 

If there is interest I'll prepare a vsgExample with my current polling solution in the main loop.

**Reply by dov** [reply]
I created a short screenrecording of the interaction of the spacemouse and the flighthelmet model. See:

https://www.youtube.com/watch?v=LxXnWlNtbSo

**Reply by smacl6301** [reply]
I've integrated my own Windows software to spacemouse using 3DxWare SDK\Lib\x64\siapp.lib and 3DxWare SDK\Lib\x64\siapp.h   See https://3dconnexion.com/us/software-developer-program/ for SDK details.

**Reply by dov** [reply]
This seems to be the way that FreeCAD does it as well if I understand the sources correctly. They use spnav for Linux and the 3DxWare SDK for windows.

**Comment by blobfish** [comment]
Hello,
I am the guy who initially added spaceball support to freecad. For those who don't know, the spaceball devices have had a sordid past. I caution anybody using the official SDK, as they like to obsolete devices much sooner than needed. Try to stick to the raw events and the devices will have much longer lives. Speaking as someone, who until recently, was still using my magellan plus xt from 1996. For all things [Spaceball](https://spacemice.org/index.php?title=Main_Page).
